diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/block/blk-core.c linux-3.2.52/block/blk-core.c
--- linux-3.2.52.orig/block/blk-core.c	2013-10-05 20:15:10.285002443 +0800
+++ linux-3.2.52/block/blk-core.c	2013-11-10 10:55:50.346080991 +0800
@@ -2516,6 +2516,10 @@ void blk_rq_bio_prep(struct request_queu
 	rq->__data_len = bio->bi_size;
 	rq->bio = rq->biotail = bio;
 
+#if defined(CONFIG_FS_TSSD)
+	rq->session_key = bio->bi_session_key;
+#endif
+
 	if (bio->bi_bdev)
 		rq->rq_disk = bio->bi_bdev->bd_disk;
 }
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/block/elevator.c linux-3.2.52/block/elevator.c
--- linux-3.2.52.orig/block/elevator.c	2013-10-05 20:15:10.285002443 +0800
+++ linux-3.2.52/block/elevator.c	2013-11-10 11:02:09.394047502 +0800
@@ -75,6 +75,14 @@ int elv_rq_merge_ok(struct request *rq,
 	if (!rq_mergeable(rq))
 		return 0;
 
+#if defined(CONFIG_FS_TSSD)
+	/* 
+	 * Don't merge two requests with different session keys
+	 */
+	if (rq->session_key != bio->bi_session_key)
+		return 0;
+#endif
+
 	/*
 	 * Don't merge file system requests and discard requests
 	 */
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/Documentation/dontdiff linux-3.2.52/Documentation/dontdiff
--- linux-3.2.52.orig/Documentation/dontdiff	2013-10-06 14:45:28.490231054 +0800
+++ linux-3.2.52/Documentation/dontdiff	2013-11-11 14:09:20.152580175 +0800
@@ -154,7 +154,6 @@ kxgettext
 lkc_defs.h
 lex.c
 lex.*.c
-linux
 logo_*.c
 logo_*_clut224.c
 logo_*_mono.c
@@ -260,3 +259,9 @@ wakeup.lds
 zImage*
 zconf.hash.c
 zoffset.h
+security
+ubuntu
+scripts
+arch
+kernel
+cayman_reg_safe.h
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/drivers/ata/libata-core.c linux-3.2.52/drivers/ata/libata-core.c
--- linux-3.2.52.orig/drivers/ata/libata-core.c	2013-10-05 20:15:11.789002913 +0800
+++ linux-3.2.52/drivers/ata/libata-core.c	2013-11-10 11:53:57.664864350 +0800
@@ -565,10 +565,10 @@ void ata_tf_to_fis(const struct ata_task
 	fis[14] = 0;
 	fis[15] = tf->ctl;
 
-	fis[16] = 0;
-	fis[17] = 0;
-	fis[18] = 0;
-	fis[19] = 0;
+	fis[16] = tf->sk0;
+	fis[17] = tf->sk1;
+	fis[18] = tf->sk2;
+	fis[19] = tf->sk3;
 }
 
 /**
@@ -843,6 +843,26 @@ int ata_build_rw_tf(struct ata_taskfile
 	return 0;
 }
 
+#if defined(CONFIG_FS_TSSD)
+/* 
+ * Build ata task file with session_key
+ */
+int ata_build_rw_tf_with_session_key(struct ata_taskfile *tf, struct ata_device *dev,
+		    u64 block, u32 n_block, unsigned int tf_flags,
+		    unsigned int tag, unsigned long session_key) {
+	int res;
+
+	res = ata_build_rw_tf(tf, dev, block, n_block, tf_flags, tag);
+	if(res) return res;
+
+	tf->sk0 = (session_key >> 0) & 0xFF;
+	tf->sk1 = (session_key >> 8) & 0xFF;
+	tf->sk2 = (session_key >> 16) & 0xFF;
+	tf->sk3 = (session_key >> 24) & 0xFF;
+	return 0;
+}
+#endif
+
 /**
  *	ata_pack_xfermask - Pack pio, mwdma and udma masks into xfer_mask
  *	@pio_mask: pio_mask
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/drivers/ata/libata.h linux-3.2.52/drivers/ata/libata.h
--- linux-3.2.52.orig/drivers/ata/libata.h	2013-10-05 20:15:11.785002911 +0800
+++ linux-3.2.52/drivers/ata/libata.h	2013-11-10 12:04:26.188643718 +0800
@@ -66,6 +66,12 @@ extern struct ata_queued_cmd *ata_qc_new
 extern int ata_build_rw_tf(struct ata_taskfile *tf, struct ata_device *dev,
 			   u64 block, u32 n_block, unsigned int tf_flags,
 			   unsigned int tag);
+#if defined(CONFIG_FS_TSSD)
+extern int ata_build_rw_tf_with_session_key(struct ata_taskfile *tf, struct ata_device *dev,
+			   u64 block, u32 n_block, unsigned int tf_flags,
+			   unsigned int tag, unsigned long session_key);
+#endif
+
 extern u64 ata_tf_read_block(struct ata_taskfile *tf, struct ata_device *dev);
 extern unsigned ata_exec_internal(struct ata_device *dev,
 				  struct ata_taskfile *tf, const u8 *cdb,
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/drivers/ata/libata-scsi.c linux-3.2.52/drivers/ata/libata-scsi.c
--- linux-3.2.52.orig/drivers/ata/libata-scsi.c	2013-10-05 20:15:11.785002911 +0800
+++ linux-3.2.52/drivers/ata/libata-scsi.c	2013-11-10 11:58:24.568760472 +0800
@@ -1699,8 +1699,14 @@ static unsigned int ata_scsi_rw_xlat(str
 	qc->flags |= ATA_QCFLAG_IO;
 	qc->nbytes = n_block * scmd->device->sector_size;
 
+#if defined(CONFIG_FS_TSSD)
+	rc = ata_build_rw_tf_with_session_key(&qc->tf, qc->dev, block, n_block,
+			tf_flags, qc->tag, 
+			scmd->request ? scmd->request->session_key : 0L); 
+#else
 	rc = ata_build_rw_tf(&qc->tf, qc->dev, block, n_block, tf_flags,
 			     qc->tag);
+#endif
 	if (likely(rc == 0))
 		return 0;
 
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/fs/direct-io.c linux-3.2.52/fs/direct-io.c
--- linux-3.2.52.orig/fs/direct-io.c	2013-10-05 20:15:15.449004056 +0800
+++ linux-3.2.52/fs/direct-io.c	2013-10-06 14:37:46.110165952 +0800
@@ -414,7 +414,9 @@ static inline void dio_bio_submit(struct
 	unsigned long flags;
 
 	bio->bi_private = dio;
-
+#ifdef CONFIG_FS_TSSD
+	bio->bi_session_key = dio->iocb->ki_filp->f_session_key;
+#endif
 	spin_lock_irqsave(&dio->bio_lock, flags);
 	dio->refcount++;
 	spin_unlock_irqrestore(&dio->bio_lock, flags);
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/fs/ext4/ext4.h linux-3.2.52/fs/ext4/ext4.h
--- linux-3.2.52.orig/fs/ext4/ext4.h	2013-10-05 20:15:15.433004051 +0800
+++ linux-3.2.52/fs/ext4/ext4.h	2013-10-05 20:38:27.337213223 +0800
@@ -577,6 +577,7 @@ struct ext4_new_group_data {
  /* note ioctl 11 reserved for filesystem-independent FIEMAP ioctl */
 #define EXT4_IOC_ALLOC_DA_BLKS		_IO('f', 12)
 #define EXT4_IOC_MOVE_EXT		_IOWR('f', 15, struct move_extent)
+#define EXT4_IOC_SETSKEY		_IOW('f', 20, unsigned long)
 
 #if defined(__KERNEL__) && defined(CONFIG_COMPAT)
 /*
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/fs/ext4/ioctl.c linux-3.2.52/fs/ext4/ioctl.c
--- linux-3.2.52.orig/fs/ext4/ioctl.c	2013-10-05 20:15:15.441004053 +0800
+++ linux-3.2.52/fs/ext4/ioctl.c	2013-10-05 20:38:27.337213223 +0800
@@ -28,6 +28,9 @@ long ext4_ioctl(struct file *filp, unsig
 	ext4_debug("cmd = %u, arg = %lu\n", cmd, arg);
 
 	switch (cmd) {
+	case EXT4_IOC_SETSKEY:
+		filp->f_session_key = arg;
+		return 0;
 	case EXT4_IOC_GETFLAGS:
 		ext4_get_inode_flags(ei);
 		flags = ei->i_flags & EXT4_FL_USER_VISIBLE;
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/fs/Kconfig linux-3.2.52/fs/Kconfig
--- linux-3.2.52.orig/fs/Kconfig	2013-10-05 20:15:15.329004018 +0800
+++ linux-3.2.52/fs/Kconfig	2013-10-05 20:38:27.337213223 +0800
@@ -10,6 +10,10 @@ source "fs/ext2/Kconfig"
 source "fs/ext3/Kconfig"
 source "fs/ext4/Kconfig"
 
+config FS_TSSD
+    bool "Enable TrustedSSD"
+    default y
+
 config FS_XIP
 # execute in place
 	bool
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/include/linux/ata.h linux-3.2.52/include/linux/ata.h
--- linux-3.2.52.orig/include/linux/ata.h	2013-10-05 20:15:09.653002246 +0800
+++ linux-3.2.52/include/linux/ata.h	2013-11-10 11:14:35.557906114 +0800
@@ -495,6 +495,14 @@ struct ata_taskfile {
 
 	u8			device;
 
+#if defined(CONFIG_FS_TSSD)
+	/* session key = sk3 << 24 | sk2 << 16 | sk1 << 8 | sk0 */
+	u8			sk0;	
+	u8			sk1;
+	u8			sk2;
+	u8			sk3;
+#endif
+
 	u8			command;	/* IO operation */
 };
 
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/include/linux/blkdev.h linux-3.2.52/include/linux/blkdev.h
--- linux-3.2.52.orig/include/linux/blkdev.h	2013-10-05 20:15:09.737002272 +0800
+++ linux-3.2.52/include/linux/blkdev.h	2013-11-10 10:44:34.294048755 +0800
@@ -164,6 +164,10 @@ struct request {
 	unsigned int timeout;
 	int retries;
 
+#if defined(CONFIG_FS_TSSD)
+	unsigned long		session_key;
+#endif
+
 	/*
 	 * completion callback.
 	 */
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/include/linux/blk_types.h linux-3.2.52/include/linux/blk_types.h
--- linux-3.2.52.orig/include/linux/blk_types.h	2013-10-05 20:15:09.677002253 +0800
+++ linux-3.2.52/include/linux/blk_types.h	2013-10-06 14:40:04.750185426 +0800
@@ -66,6 +66,9 @@ struct bio {
 	bio_end_io_t		*bi_end_io;
 
 	void			*bi_private;
+#if defined(CONFIG_FS_TSSD)
+	unsigned long		bi_session_key;
+#endif
 #if defined(CONFIG_BLK_DEV_INTEGRITY)
 	struct bio_integrity_payload *bi_integrity;  /* data integrity */
 #endif
diff -uprN '--exclude-from=linux-3.2.52/Documentation/dontdiff' linux-3.2.52.orig/include/linux/fs.h linux-3.2.52/include/linux/fs.h
--- linux-3.2.52.orig/include/linux/fs.h	2013-10-05 20:15:09.709002263 +0800
+++ linux-3.2.52/include/linux/fs.h	2013-10-06 14:38:26.402171609 +0800
@@ -984,7 +984,9 @@ struct file {
 #define f_dentry	f_path.dentry
 #define f_vfsmnt	f_path.mnt
 	const struct file_operations	*f_op;
-
+#ifdef CONFIG_FS_TSSD
+	unsigned long		f_session_key;
+#endif
 	/*
 	 * Protects f_ep_links, f_flags, f_pos vs i_size in lseek SEEK_CUR.
 	 * Must not be taken from IRQ context.
